src/Bicep.LangServer/Completions/ModuleReferenceCompletionProvider.cs

```diff
@@ -0,0 +1,543 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Bicep.Core.Configuration;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.LanguageServer.Providers;
+using Bicep.LanguageServer.Settings;
+using Bicep.LanguageServer.Telemetry;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Bicep.LanguageServer.Completions
+{
+    public class ModuleReferenceCompletionProvider : IModuleReferenceCompletionProvider
+    {
+        private readonly IAzureContainerRegistryNamesProvider azureContainerRegistryNamesProvider;
+        private readonly IConfigurationManager configurationManager;
+        private readonly IModulesMetadataProvider modulesMetadataProvider;
+        private readonly ISettingsProvider settingsProvider;
+        private readonly ITelemetryProvider telemetryProvider;
+
+        private static readonly Dictionary<string, string> BicepRegistryAndTemplateSpecShemaCompletionLabelsWithDetails = new Dictionary<string, string>()
+        {
+            {"br:", "Bicep registry schema name" },
+            {"br/", "Bicep registry schema name" },
+            {"ts:", "Template spec schema name" },
+            {"ts/", "Template spec schema name" },
+        };
+
+        private static readonly Regex ACRModuleRegistryWithoutAlias = new Regex(@"br:(?<registry>(.*).azurecr.io)/", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
```

Comment:
>azurecr [](http://example.com/codeflow?start=102&length=7)

We can't assume azurecr.io for ACR. They are working on a public preview feature to enable custom domains: [https://github.com/Azure/acr/blob/main/docs/custom-domain/README.md](https://github.com/Azure/acr/blob/main/docs/custom-domain/README.md)